datalad/distribution/tests/test_uninstall.py

```diff
@@ -98,23 +143,147 @@ def test_uninstall_subdataset(src, dst):
+    ok_clean_git(ds.path)
+    files_left = glob(opj(ds.path, '*', '*', '*')) + glob(opj(ds.path, '*'))
+    #import pdb; pdb.set_trace()
+    ok_(all([f.endswith('keep') for f in files_left if exists(f) and not isdir(f)]))
+    ok_(all([f.endswith('kill') for f in files_left
+             if not exists(f) and not isdir(f)]))
+    # drop handles for all 'kill' files
+    ds.uninstall(['kill', opj('deep', 'dir', 'kill')], recursive=True, check=False,
+                 remove_handles=True)
+    ok_clean_git(ds.path)
+    files_left = glob(opj(ds.path, '*', '*', '*')) + glob(opj(ds.path, '*'))
+    ok_(all([f.endswith('keep') for f in files_left if exists(f) and not isdir(f)]))
+    ok_(not any([f.endswith('kill') for f in files_left]))
+
+
+@with_tempfile()
+def test_uninstall_dataset(path):
+    ds = Dataset(path)
+    ok_(not ds.is_installed())
+    ds.create()
+    ok_(ds.is_installed())
+    ok_clean_git(ds.path)
+    # would only drop data
+    ds.uninstall()
+    # actually same as this, for cmdline compat reasons
+    ds.uninstall(path=[])
+    ok_clean_git(ds.path)
+    # removing all handles equal removal of entire dataset, needs safety switch
+    assert_raises(ValueError, ds.uninstall, remove_handles=True)
+    ds.uninstall(remove_handles=True, remove_history=True)
+    # completely gone
+    ok_(not ds.is_installed())
+    ok_(not exists(ds.path))
+
 
+@with_tree({'one': 'test', 'two': 'test'})
+def test_remove_file_handle_only(path):
+    ds = Dataset(path).create(force=True)
+    ds.add(os.curdir)
+    ok_clean_git(ds.path)
+    # make sure there is any key
+    ok_(len(ds.repo.get_file_key('one')))
+    # both files link to the same key
+    eq_(ds.repo.get_file_key('one'),
+        ds.repo.get_file_key('two'))
+    rpath_one = realpath(opj(ds.path, 'one'))
+    eq_(rpath_one, realpath(opj(ds.path, 'two')))
+    path_two = opj(ds.path, 'two')
+    ok_(exists(path_two))
+    # remove one handle, should not affect the other
+    ds.uninstall('two', remove_data=False, remove_handles=True)
+    eq_(rpath_one, realpath(opj(ds.path, 'one')))
+    ok_(exists(rpath_one))
+    ok_(not exists(path_two))
 
-def test_uninstall_dataset():
-    raise SkipTest("TODO")
 
+@with_tree({'deep': {'dir': {'test': 'testcontent'}}})
+def test_uninstall_recursive(path):
+    ds = Dataset(path).create(force=True)
+    subds = ds.create_subdataset('deep', force=True)
+    # we add one file
+    eq_(len(subds.add('.')), 1)
+    # save all -> all clean
+    ds.save(auto_add_changes=True, recursive=True)
+    ok_clean_git(subds.path)
+    ok_clean_git(ds.path)
+    # now uninstall in subdataset through superdataset
+    target_fname = opj('deep', 'dir', 'test')
+    # sane starting point
+    ok_(exists(opj(ds.path, target_fname)))
+    # uninstallation fails without --recursive
+    assert_raises(ValueError, ds.uninstall, target_fname)
+    # doesn't have the minimum number of copies for a safe drop
+    # TODO: better exception
```

Comment:
True. In general, we should stop adding those todos altogether and replace them with issues.